src/H5Dmpio.c

```diff
@@ -5601,256 +5350,128 @@ H5D__mpio_get_chunk_insert_info_types(MPI_Datatype *contig_type, hbool_t *contig
+ *              structures for filtered chunks, populates I/O vectors and
+ *              performs vector I/O on those chunks.
  *
  * Return:      Non-negative on success/Negative on failure
  *
  *-------------------------------------------------------------------------
  */
 static herr_t
-H5D__mpio_collective_filtered_io_type(H5D_filtered_collective_io_info_t *chunk_list, size_t num_entries,
-                                      H5D_io_op_type_t op_type, MPI_Datatype *new_mem_type,
-                                      hbool_t *mem_type_derived, MPI_Datatype *new_file_type,
-                                      hbool_t *file_type_derived)
+H5D__mpio_collective_filtered_vec_io(H5F_shared_t *f_sh, H5D_filtered_collective_io_info_t *chunk_list,
+                                     size_t num_entries, H5D_io_op_type_t op_type)
 {
-    MPI_Aint *io_buf_array      = NULL; /* Relative displacements of filtered chunk data buffers */
-    MPI_Aint *file_offset_array = NULL; /* Chunk offsets in the file */
-    int      *length_array      = NULL; /* Filtered Chunk lengths */
-    int       mpi_code;
-    herr_t    ret_value = SUCCEED;
+    const void **io_wbufs    = NULL;
+    void       **io_rbufs    = NULL;
+    H5FD_mem_t  *io_types    = NULL;
+    uint32_t     iovec_count = 0;
+    haddr_t     *io_addrs    = NULL;
+    size_t      *io_sizes    = NULL;
+    herr_t       ret_value   = SUCCEED;
 
     FUNC_ENTER_PACKAGE
 
-    assert(chunk_list || 0 == num_entries);
-    assert(new_mem_type);
-    assert(mem_type_derived);
-    assert(new_file_type);
-    assert(file_type_derived);
-
-    *mem_type_derived  = FALSE;
-    *file_type_derived = FALSE;
-    *new_mem_type      = MPI_BYTE;
-    *new_file_type     = MPI_BYTE;
+    assert(f_sh);
+    assert(chunk_list || num_entries == 0);
 
     if (num_entries > 0) {
-        H5F_block_t *chunk_block;
-        size_t       last_valid_idx = 0;
-        size_t       i;
-        int          chunk_count;
-
-        /*
-         * Determine number of chunks for I/O operation and
-         * setup for derived datatype creation if I/O operation
-         * includes multiple chunks
-         */
-        if (num_entries == 1) {
-            /* Set last valid index to 0 for contiguous datatype creation */
-            last_valid_idx = 0;
+        if (num_entries > UINT32_MAX)
+            HGOTO_ERROR(H5E_INTERNAL, H5E_BADRANGE, FAIL,
+                        "number of chunk entries in I/O operation exceeds UINT32_MAX")
 
-            if (op_type == H5D_IO_OP_WRITE)
-                chunk_count = 1;
-            else
-                chunk_count = chunk_list[0].need_read ? 1 : 0;
-        }
+        if (op_type == H5D_IO_OP_WRITE)
+            iovec_count = (uint32_t)num_entries;
         else {
-            MPI_Aint chunk_buf;
-            MPI_Aint base_buf;
-            haddr_t  base_offset = HADDR_UNDEF;
+            /* For read operations, determine number of chunks actually
```

Comment:
Couldn't this loop be avoided by tracking the # of chunks being read while the list was being created?  (i.e. when the 'need_read' flag is set on each chunk)

Reply:
Sure, this should be fairly easy to implement. Though, there are a few places where this field may get updated after the list is constructed that would need to be accounted for, so this was the simplest way to deal with that.

Reply:
@qkoziol After thinking about this for a bit, it looks like the most reasonable way to handle this would be to add a `size_t` hint parameter ("no hint provided" represented as `SIZE_MAX`) to this function that lets the caller specify an expected number of entries for the I/O vectors and then check and grow the buffers if the caller was wrong with their hint. It's a little clunky, but shouldn't be bad to implement if it seems worth doing over another scan through the chunk list.

The issue is that in the current implementation there is no "top-level" data structure that gets passed between internal functions which could hold the expected number of entries. The main data structure is an array of one struct per chunk that has to be iterated through to determine how many chunks will be involved in the I/O.

Reply:
I would suggest creating that 'top-level' struct, then you can put this counter in it, as well as a pointer to the array of chunks.  Then the # of parameters will stay the same, but you can pass around more information.